Un/-block consumer groups in data migrations

bashtanov opened the pull request:
Integrate consumer groups blocking into migration reconciliation process.

Add consumer groups topic into the list of topics migrated, and schedule relevant work on its partitions holding affected groups in the same way as we do it with data partitions. Dumping restoring of offsets will be implemented via different API in a subsequent PR.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
app: proxy class to operate kafka groups from cluster data migrations

Cluster layer cannot depend on kafka layer due to circular dependencies,
so we'll need to use a proxy class with virtual functions.

Seastar's `sharded<Service>` template cannot be declared with base type
but initialized using a constructor of a derived type. It is a bit
annoying, as `ss::shared_ptr<Service>` used inside to store the
instances actually can be initialized with a
`ss::shared_ptr<ServiceSubclass> &&`, but that's not exposed in
`sharded`. Use pimpl when wiring, but pass actual proxy object
references to caller class instances.

bashtanov committed:
tests/data_migrate: smoke-test consumer groups migrations

bashtanov committed:
c/data-migrate/backend: refactoring

1) allow lookups by topic_namsepace_view in topic_map_t
2) when retrieving per-topic parameters from migration definition for to
use in work info objects, separate getting inbound topic by index to a
function
3) make work info object creation functions const
4) avoid excessive copying when spawning topic work
5) defensively use .at() instead of operator[] where the entry is
knowingly present in the collection
6) simplify handle_migration_update() to use optional::transform

bashtanov committed:
c/data-migrate/backend: populate topic work structures with group data

bashtanov committed:
tests/data-migrate/api: stop producer early on failure

bashtanov committed:
k/s/group_manger: store blocked groups in partitions

This will ensure correct recovery, as it is partition-wise

bashtanov committed:
k/s/group_manager: add function to delete a group

bashtanov committed:
c/data-migrate/backend: partition work scope rearrangements

1. Decide whether partition-scoped work is needed based on the topic.
It is needed because in the final stage of an outbound migrations data
topics are removed so cannot do even no-op partition work
(partitions are gone), while consumer offset topic needs to delete and
unblock groups, which is partition scoped work.
2. Allow stages with neither partition- nor topic-scoped work

bashtanov committed:
tests/data-migrate: validate group migrations (blocking only for now)

bashtanov committed:
c/data-migrate/worker: implement group un/-blocking for outbound

bashtanov committed:
c/data_migrate: include __consumer_offsets topic in migration

Currently migrations operate data topics only. If consumer groups were
scheduled to be migrated include the topic they are stored in will be
included in the reconciliation process.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68520-->
<!--0dc220f4-c831-4a38-ab31-f04ae09beeab-->
<!--title-->
## Retry command for [Build#68520](https://buildkite.com/redpanda/redpanda/builds/68520)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_creating_and_listing_migrations
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--1d1accf2-3607-4a10-80e7-b6e999ea1b7a-->
<!--TEST RESULTS-->
<!--ed20ef5b-8ad3-40a0-a4d9-30dca63ac578-->
<!--TEST RESULTS-->
<!--3d50675c-d283-43fb-902e-5ff559e2f674-->
<!--TEST RESULTS-->
<!--09b93fd9-94a5-4565-8c0f-093005b4e96b-->
<!--TEST RESULTS-->
<!--212fbaa4-48f6-4d41-b3cc-6aa77e815115-->
<!--TEST RESULTS-->
<!--2887dbbb-4f62-48de-aacd-dad51dce8c93-->
<!--TEST RESULTS-->
<!--dc996406-a86d-4785-af83-ce3f8e2251a2-->
<!--TEST RESULTS-->
<!--1518485e-85a8-4101-85a5-d966c665375e-->
<!--TEST RESULTS-->
<!--1ac88a18-616f-4b20-b470-2314c8300571-->
<!--TEST RESULTS-->
<!--cb211c28-cf74-499a-9651-3a37d9a88956-->
<!--TEST RESULTS-->
<!--cf0ccc0a-c54f-4a5c-95dc-9513bc7aaed0-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68520">build#68520</a></summary>


| test_class                   | test_method                          | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-----------------------------|:-------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest | test_cloud_storage                   | {"cleanup_policy": "compact,delete"}                                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68520#0197e52b-3e1e-4a72-ae2b-e04738b54004 | FLAKY         | 19/21    | upstream reliability is '96.10778443113772'. current run reliability is '90.47619047619048'. drift is 5.63159 and the allowed drift is set to 50. The test should PASS  |
| DataMigrationsApiTest        | test_creating_and_listing_migrations |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68520#0197e52b-3e20-49d7-85ea-a3c45c09818f | FLAKY         | 12/21    | upstream reliability is '100.0'. current run reliability is '57.14285714285714'. drift is 42.85714 and the allowed drift is set to 50. The test should PASS             |
| DataMigrationsApiTest        | test_creating_and_listing_migrations |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68520#0197e52b-d0a4-4cac-ad74-ecff9b98ff0d | FLAKY         | 10/21    | upstream reliability is '100.0'. current run reliability is '47.61904761904761'. drift is 52.38095 and the allowed drift is set to 50. The test should FAIL             |
| DataMigrationsApiTest        | test_higher_level_migration_api      |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68520#0197e52b-d0a0-4d08-aac9-85218138abe7 | FLAKY         | 11/21    | upstream reliability is '100.0'. current run reliability is '52.38095238095239'. drift is 47.61905 and the allowed drift is set to 50. The test should PASS             |
| DatalakeMetricsTest          | test_lag_metrics                     | {"cloud_storage_type": 1}                                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68520#0197e52b-3e1e-44d3-99a0-4be960a93c0f | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| RandomNodeOperationsTest     | test_node_operations                 | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68520#0197e52b-3e20-49f2-8b83-64b174576772 | FLAKY         | 20/21    | upstream reliability is '99.10313901345292'. current run reliability is '95.23809523809523'. drift is 3.86504 and the allowed drift is set to 50. The test should PASS  |
| DisablingPartitionsTest      | test_disable                         |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68520#0197e52b-d0a0-4d08-aac9-85218138abe7 | FLAKY         | 16/21    | upstream reliability is '89.81818181818181'. current run reliability is '76.19047619047619'. drift is 13.62771 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68731">build#68731</a></summary>


| test_class            | test_method                     | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      |
|:----------------------|:--------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest | test_higher_level_migration_api |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68731#0197f4c5-2134-444d-bfdd-af3b77fe8cd0 | FLAKY         | 14/21    | upstream reliability is '100.0'. current run reliability is '66.66666666666666'. drift is 33.33333 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68814">build#68814</a></summary>


| test_class                  | test_method                              | test_arguments                                                                                                                | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:-----------------------------------------|:------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerOffsetsRecoveryTest | test_consumer_offsets_partition_recovery |                                                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68814#0197f8e0-4b4d-4235-84be-8befeed9df3b | FLAKY         | 20/21    | upstream reliability is '98.61431870669746'. current run reliability is '95.23809523809523'. drift is 3.37622 and the allowed drift is set to 50. The test should PASS |
| DataMigrationsApiTest       | test_higher_level_migration_api          |                                                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68814#0197f8e0-b7b9-4a76-8084-ccb4e057f777 | FLAKY         | 10/21    | upstream reliability is '100.0'. current run reliability is '47.61904761904761'. drift is 52.38095 and the allowed drift is set to 50. The test should FAIL            |
| TieredStorageTest           | test_tiered_storage                      | {"cloud_storage_type_and_url_style": [1, "path"], "test_case": {"name": "(TS_Read == True, SegmentRolledByTimeout == True)"}} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68814#0197f8e0-b7b9-4a76-8084-ccb4e057f777 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68926">build#68926</a></summary>


| test_class                | test_method                        | test_arguments                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:-----------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest | test_read_chunks                   |                                                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68926#01980a13-83d9-481a-b4f4-23b557de933c | FLAKY         | 20/21    | upstream reliability is '98.63760217983652'. current run reliability is '95.23809523809523'. drift is 3.39951 and the allowed drift is set to 50. The test should PASS |
| DataMigrationsApiTest     | test_migrated_topic_data_integrity | {"include_groups": true, "params": {"cancellation": {"dir": "in", "stage": "prepared"}, "use_alias": true}, "transfer_leadership": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68926#01980a12-949a-408d-9664-ed3c87aa52f3 | FLAKY         | 20/21    | upstream reliability is 'nan'. current run reliability is '95.23809523809523'. drift is nan and the allowed drift is set to 50. The test should FAIL                   |
| LogCompactionTest         | compaction_stress_test             | {"cleanup_policy": "compact", "key_set_cardinality": 100, "storage_compaction_key_map_memory_kb": 131072}                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68926#01980a12-949a-408d-9664-ed3c87aa52f3 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| NodePostRestartProbeTest  | post_restart_probe_test            |                                                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68926#01980a12-949d-4de8-ba5b-b45ba6fa0bf2 | FLAKY         | 19/21    | upstream reliability is '99.52830188679245'. current run reliability is '90.47619047619048'. drift is 9.05211 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69200">build#69200</a></summary>


| test_class                              | test_method                                      | test_arguments                                                                                                                            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------------------|:-------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerGroupBalancingTest              | test_coordinator_nodes_balance                   |                                                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69200#01981976-4319-418c-978f-5fb0857b1c8a | FLAKY         | 19/21    | upstream reliability is '99.71264367816092'. current run reliability is '90.47619047619048'. drift is 9.23645 and the allowed drift is set to 50. The test should PASS  |
| DataMigrationsApiTest                   | test_migrated_topic_data_integrity               | {"include_groups": true, "params": {"cancellation": null, "use_alias": true}, "transfer_leadership": true}                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69200#01981979-0850-40df-8de6-9a2c33caae58 | FLAKY         | 19/21    | The test was found to be new, and no failures are allowed                                                                                                               |
| DataMigrationsApiTest                   | test_migrated_topic_data_integrity               | {"include_groups": true, "params": {"cancellation": {"dir": "in", "stage": "preparing"}, "use_alias": true}, "transfer_leadership": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69200#01981976-4319-418c-978f-5fb0857b1c8a | FLAKY         | 20/21    | upstream reliability is 'nan'. current run reliability is '95.23809523809523'. drift is nan and the allowed drift is set to 50. The test should FAIL                    |
| InternalTopicProtectionLargeClusterTest | test_consumer_offset_topic                       |                                                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69200#01981979-085f-433f-93e2-bde819e2a0ed | FLAKY         | 20/21    | upstream reliability is '99.71264367816092'. current run reliability is '95.23809523809523'. drift is 4.47455 and the allowed drift is set to 50. The test should PASS  |
| ShadowIndexingLocalRetentionTest        | test_shadow_indexing_non_default_local_retention | {"cloud_storage_type": 2}                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69200#01981976-4318-4205-bbe4-f6b15d32c9de | FLAKY         | 20/21    | upstream reliability is '87.48180494905385'. current run reliability is '95.23809523809523'. drift is -7.75629 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69262">build#69262</a></summary>


| test_class               | test_method             | test_arguments                                                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArchivalTest             | test_connection_flicker | {"cloud_storage_type": 2}                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69262#01981d19-8916-4ebc-8189-76555910fc31 | FLAKY         | 20/21    | upstream reliability is '99.23076923076923'. current run reliability is '95.23809523809523'. drift is 3.99267 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations    | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69262#01981d16-75ea-4f1c-8ff7-996ad7ef662f | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest | test_node_operations    | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69262#01981d16-75f0-4d27-934d-d733c743eaa0 | FLAKY         | 19/21    | upstream reliability is '99.2831541218638'. current run reliability is '90.47619047619048'. drift is 8.80696 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69275">build#69275</a></summary>


| test_class               | test_method                   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-------------------------|:------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| FeaturesMultiNodeTest    | test_license_upload_and_query |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69275#01981e0e-4e78-4b22-90e6-62bb3087fe9c | FLAKY         | 19/21    | upstream reliability is '98.4126984126984'. current run reliability is '90.47619047619048'. drift is 7.93651 and the allowed drift is set to 50. The test should PASS |
| NodePostRestartProbeTest | post_restart_probe_test       |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69275#01981e0e-fce4-489d-82e5-f3be00ab3017 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS             |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69290">build#69290</a></summary>


| test_class               | test_method                          | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------------|:-------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest    | test_creating_and_listing_migrations |                                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69290#01981f0a-7299-4e06-926a-cf5eb5842669 | FLAKY         | 18/21    | upstream reliability is '99.52267303102626'. current run reliability is '85.71428571428571'. drift is 13.80839 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations                 | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69290#01981f08-3025-4c30-85f1-345311516a64 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69311">build#69311</a></summary>


| test_class                     | test_method                             | test_arguments                                                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   |
|:-------------------------------|:----------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageScrubberTest       | test_scrubber                           | {"cloud_storage_type": 1}                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69311#019827e5-7521-4935-8ab7-bd1bfb75d5bc | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS                |
| CloudStorageTimingStressTest   | test_cloud_storage_with_partition_moves | {"cleanup_policy": "compact,delete"}                                                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69311#019827e5-e181-41c9-863d-5cc77240360c | FLAKY         | 20/21    | upstream reliability is '99.48320413436691'. current run reliability is '95.23809523809523'. drift is 4.24511 and the allowed drift is set to 50. The test should PASS   |
| DatalakeCustomPartitioningTest | test_sticky_default                     | {"catalog_type": "nessie", "cloud_storage_type": 1}                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69311#019827e5-e17d-4f10-b7fc-b9a330488721 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS                |
| RandomNodeOperationsTest       | test_node_operations                    | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69311#019827e5-e182-497f-821c-7ba834764609 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS               |
| RandomNodeOperationsTest       | test_node_operations                    | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69311#019827e5-e17d-4f10-b7fc-b9a330488721 | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest       | test_node_operations                    | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69311#019827e5-e17e-4655-9a18-375aca408ffb | FLAKY         | 10/21    | upstream reliability is '99.3103448275862'. current run reliability is '47.61904761904761'. drift is 51.6913 and the allowed drift is set to 50. The test should FAIL    |
| RandomNodeOperationsTest       | test_node_operations                    | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69311#019827e5-e17f-4baa-8c80-4e137c889ab2 | FLAKY         | 13/21    | upstream reliability is '99.66101694915255'. current run reliability is '61.904761904761905'. drift is 37.75626 and the allowed drift is set to 50. The test should PASS |
| SimpleEndToEndTest             | test_relaxed_acks                       | {"write_caching": false}                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69311#019827e5-e17e-40ab-8463-2034fe8e40e8 | FLAKY         | 20/21    | upstream reliability is '99.58847736625515'. current run reliability is '95.23809523809523'. drift is 4.35038 and the allowed drift is set to 50. The test should PASS   |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69349">build#69349</a></summary>


| test_class               | test_method                  | test_arguments                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------------|:-----------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations         | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69349#01982d93-1f11-4302-83ba-e3f1b1a4d549 | FLAKY         | 15/21    | upstream reliability is '99.61832061068702'. current run reliability is '71.42857142857143'. drift is 28.18975 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations         | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69349#01982d93-1f0f-42da-98a2-d253d3dd0c12 | FLAKY         | 19/21    | upstream reliability is '99.57446808510639'. current run reliability is '90.47619047619048'. drift is 9.09828 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest | test_node_operations         | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69349#01982d93-1f12-4ea9-b25b-de0c72f5afaa | FLAKY         | 20/21    | upstream reliability is '99.47089947089947'. current run reliability is '95.23809523809523'. drift is 4.2328 and the allowed drift is set to 50. The test should PASS   |
| RandomNodeOperationsTest | test_node_operations         | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69349#01982d93-1f0e-4dbc-8ec3-b7aed8cddfb2 | FLAKY         | 19/21    | upstream reliability is '97.84366576819407'. current run reliability is '90.47619047619048'. drift is 7.36748 and the allowed drift is set to 50. The test should PASS  |
| RpkACLTest               | test_modify_then_query_error |                                                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69349#01982d93-1f10-49f2-b9d3-926c5719f246 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| SimpleEndToEndTest       | test_relaxed_acks            | {"write_caching": false}                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69349#01982d93-1f0d-4d57-906f-61b59b758101 | FLAKY         | 20/21    | upstream reliability is '99.18032786885246'. current run reliability is '95.23809523809523'. drift is 3.94223 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69506">build#69506</a></summary>


| test_class                        | test_method             | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------------|:------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| cluster_metadata_uploader_fixture | test_upload_in_term     |                                                                                                                                         | unit        | https://buildkite.com/redpanda/redpanda/builds/69506#01983660-35b5-42b4-a6a3-1633002ccf74 | FAIL          | 0/1      |                                                                                                                                                                         |
| ArchivalTest                      | test_connection_flicker | {"cloud_storage_type": 2}                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/69506#019836f0-0326-4228-979c-45632753b3d5 | FLAKY         | 20/21    | upstream reliability is '99.22178988326849'. current run reliability is '95.23809523809523'. drift is 3.98369 and the allowed drift is set to 50. The test should PASS  |
| EndToEndCloudTopicsTxTest         | test_write              | null                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/69506#0198368e-4d5a-4606-a072-50b98b2b0651 | FLAKY         | 17/21    | upstream reliability is '95.71045576407506'. current run reliability is '80.95238095238095'. drift is 14.75807 and the allowed drift is set to 50. The test should PASS |
| ConsumerGroupTest                 | test_group_lag_metrics  | null                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/69506#0198368e-4d57-4b4f-8bf5-9b7e432b1f1e | FLAKY         | 20/21    | upstream reliability is '98.87640449438202'. current run reliability is '95.23809523809523'. drift is 3.63831 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest          | test_node_operations    | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/69506#0198368e-4d58-4400-9909-ea0bf2a52ce1 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              |
</details>
<!--/body-->

joe-redpanda commented on src/v/cluster/data_migration_backend.cc:
```diff
@@ -1319,6 +1324,28 @@ ss::future<> backend::reconcile_migration(
       mrstate.scope.sought_state);
+
+    if (!mrstate.partition_group_map) {
+        while (!co_await _group_proxy.assure_topic_exists(
```
is there a retry utility for this?

bashtanov replied:
I'm not aware of one. Other uses of `group_initializer::assure_topic_exists` can return an error, but migrations machinery retries indefinitely.

joe-redpanda commented on src/v/cluster/data_migration_backend.cc:
```diff
@@ -1592,13 +1655,13 @@ backend::work_scope backend::get_work_scope(
     switch (metadata.state) {
     case state::preparing:
-        return {state::prepared, true, false};
+        return {state::prepared, true, false, false};
```
Can we do designated initialization on these

joe-redpanda commented on src/v/kafka/server/data_migration_group_proxy_impl.cc:
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2020 Redpanda Data, Inc.
```
2025?

joe-redpanda commented on tests/rptest/tests/data_migrations_api_test.py:
```diff
@@ -717,67 +726,88 @@ def start_consumer(self, topic):
 
         assert expected_to_pass == success, f"Operation {op_name} outcome is not " \
-            f"expected. Expected to pass: {expected_to_pass}, succeeded: {success}"
+            f"expected. {expected_to_pass=}, {success=}, {result=}"
```
thats a cool feature

bharathv commented on src/v/cluster/data_migration_group_proxy.h:
```diff
@@ -0,0 +1,45 @@
+      bool to_block)
+      = 0;
+
+    virtual ss::future<std::error_code> delete_groups(
```
nit: probably worth documenting what ntp this is

bashtanov replied:
I'll change its name to `co_ntp` like above.

bharathv commented on src/v/kafka/server/group.cc:
```diff
@@ -1474,6 +1474,14 @@ void group::remove_member(member_ptr member) {
     }
 }
 
+void group::remove_full_members() {
+    while (!_members.empty()) {
+        auto member = _members.begin()->second;
+        member->expire_timer().cancel();
+        remove_member(member);
+    }
+}
```
what about pending members?  

unrelated to this PR.. I have a hard time to trying to reason about how member state changes are persisted 🤔

bashtanov replied:
> what about pending members?

They don't seem to block group deletion

> unrelated to this PR.. I have a hard time to trying to reason about how member state changes are persisted 🤔

Yeah the one where we ignore replication result scares me the most.

Maybe instead of removing members one-by-one and waiting for the rebalance we should just force-delete the group and replicate a tombstone for it regardless its members status. I should have asked about it in the joint discussion yesterday.

bharathv reviewed (commented):
I'm not super familiar with data_migraion*.h/cc.. would prefer @mmaslankaprv  to review those.

bashtanov committed:
c/data-migrate/backend: in get_work_scope use designated initializers

copilot-generated change

mmaslankaprv commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1802,6 +1804,35 @@ group_manager::describe_partition_producers(const model::ntp& ntp) {
+        g->remove_full_members();
+    });
+
+    std::vector<std::pair<model::ntp, group_id>> groups_with_ntps{
```
should we use chunked vector, or is this intended to be small ? i guess per group vector may grow pretty large

mmaslankaprv commented on src/v/cluster/data_migration_group_proxy.h:
```diff
@@ -0,0 +1,45 @@
+ * Abstract class for a proxy component to operate kafka layer consumer groups
+ * from data migration components.
+ */
+class group_proxy {
```
i am wondering if `group_proxy` could leverage the pimpl itself, the the `std::unique_ptr` trick wouldn't be required in sharded but rather than that we would use a `make_group_proxy` method

bashtanov replied:
As discussed on slack, I actually prefer this "trick": why create our own wrapper class if `unique_ptr` does exactly what we need? In particular, I like the idea of passing a reference to implementation itself to dependent components, to avoid extra indirections when calling. That's to improve both readability and performance.

mmaslankaprv commented on src/v/cluster/data_migration_backend.h:
```diff
@@ -52,24 +57,43 @@ class backend {
+        bool any_partition_work_needed() {
+            return data_partition_work_needed || co_partition_work_needed;
+        };
+        bool partition_work_needed(const model::topic_namespace& nt) {
```
nit: can be const method

mmaslankaprv commented on src/v/cluster/data_migration_types.h:
```diff
@@ -290,9 +308,13 @@ data_migration copy_migration(const data_migration& migration);
     std::optional<model::topic_namespace> source;
     std::optional<cloud_storage_location> cloud_storage_location;
+    chunked_vector<consumer_group>
+      groups; // not empty iff partition of consumer offsets topic
```
nit: typo

mmaslankaprv commented on src/v/cluster/data_migration_types.h:
```diff
@@ -290,9 +308,13 @@ data_migration copy_migration(const data_migration& migration);
 struct inbound_partition_work_info {
```
I am wondering if it is possible to add a comment that it is either a work related with consumer groups or data partition. It is now pretty confusing to see the pattern: 

```
if(groups.empty()) {
    do...
    return;
} else {
    do...
    return;
}

```
This implies the mutual exclusion of group existance and standard partition work. 

I am wondering if we could use `std::variant` to encode that information in the typesystem

bashtanov replied:
The comment is a couple of lines below, `iff` is not a typo, it means `if and only if`.

An std::variant would probably look nice, but will add rigidity so that we cannot add a trait common to both data and consumer group partitions. We have this problem with in- and outbound migrations, and it's annoying to apply visitors every so often. Here it's less likely, so I'm on the verge TBH. If you feel strongly about the visitor let me know and I'll change it.

mmaslankaprv replied:
would it be possible not to use `iff` acronym or include the full version in the parenthesis when it is introduced for the first time ?

bashtanov replied:
This word (in lowercase) became part of English language I'm afraid. It is used in 37 comments in the repo already and I only wrote 3 of them. It is used ~50 times on the company slack by various authors, only one is from me. It is present in a few reputable online dictionaries too.

bashtanov committed:
k/s/group_{manager,router}: group delete internals use chunked vector

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68814-->
<!--55375524-774e-4312-aa9e-b968b72abaf7-->
<!--title-->
## Retry command for [Build#68814](https://buildkite.com/redpanda/redpanda/builds/68814)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_higher_level_migration_api
```
<!--/body-->

bashtanov committed:
c/data-migrate/backend: Re-step-up as coordinator on term change.

We need to restart coordinating if another node have been a leader and
potentially a coordinator between our coordinatorship terms.
This is to recollect metadata of affected topics. While
data_migrated_resources guards from topic metadata changes in presence
of an active migration, another coordinator may have start or complete
migrations between our terms.

mmaslankaprv commented:
@bashtanov it seems that the tests failures are related

bashtanov commented:
yes I'm working on them

bashtanov committed:
c/data-migrate/fe: on migration creation create __consumer_offsets topic

... if it is not there.

We cannot easily do it in backend, as some reconciliation may happen on
startup when controller is leaderless and thus unable to create topics.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68926-->
<!--2f3f7080-eaa5-4f85-8d4b-ef95a732f20c-->
<!--title-->
## Retry command for [Build#68926](https://buildkite.com/redpanda/redpanda/builds/68926)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"include_groups":true,"params":{"cancellation":{"dir":"in","stage":"prepared"},"use_alias":true},"transfer_leadership":true}
```
<!--/body-->

bashtanov commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69200-->
<!--967ac6e6-51db-4bfb-83f3-307b1a3db0f2-->
<!--445779cb-7791-4cf5-ad69-f0ec607421ba-->
<!--title-->
## Retry command for [Build#69200](https://buildkite.com/redpanda/redpanda/builds/69200)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"include_groups":true,"params":{"cancellation":null,"use_alias":true},"transfer_leadership":true}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"include_groups":true,"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":true},"transfer_leadership":true}
```
<!--/body-->

bashtanov commented:
/dt

bashtanov commented:
/dt

bashtanov commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69311-->
<!--1eece7ea-2448-4941-b30a-dfc80f9580c4-->
<!--title-->
## Retry command for [Build#69311](https://buildkite.com/redpanda/redpanda/builds/69311)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"chunked_sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
```
<!--/body-->

bashtanov commented:
@mmaslankaprv I would appreciate if you could review. Failures look unrelated to me.

mmaslankaprv commented on tests/rptest/tests/data_migrations_api_test.py:
```diff
@@ -717,67 +730,88 @@ def start_consumer(self, topic):
-            operation=lambda topic: rpk.consume(topic=topic, n=1, offset=0))
+            operation=lambda topic, _: rpk.consume(topic=topic, n=1, offset=0))
+
+        if group is not None:
+
+            def read_with_group(topic, group):
+                with self.ck_consumer(group) as consumer:
+                    consumer.subscribe([topic])
+                    msg = consumer.poll(20)
+                    if msg is None or msg.error() is not None:
+                        raise ck.KafkaException(
+                            f"Failed to read from topic {topic} with group {group}: {msg and msg.error()}"
+                        )
+
+            self._do_validate_operation(**entities,
+                                        op_name="read_with_group",
```
I am wondering if it would be more robust to try to commit the offset here explicitly not having to rely on consumer on the consumer behavior. But OTOH the consumer is doing Join group, so it is good to test that

bashtanov replied:
Yes, I tried it but had to switch to reading because of needing to join first.

mmaslankaprv commented on src/v/cluster/data_migration_frontend.cc:
```diff
@@ -253,6 +255,20 @@ ss::future<check_ntp_states_reply> frontend::check_ntp_states_on_foreign_node(
+          [](const auto& migration) { return !empty(migration.groups); },
+          migration)) {
+        if (!co_await _group_proxy.assure_topic_exists(
+              model::timeout_clock::now() + 5s)) {
```
should that use `_operation_timeout` ?

bashtanov commented:
[The failure is unrelated](https://redpandadata.slack.com/archives/C06MCRY75U7/p1753180286777019). @mmaslankaprv could you merge it please?

bashtanov commented:
/ci-repeat 1
